File: learnLeetCode/Day29/Part2/monster_survival.py
from collections import deque
from typing import List, Tuple

def max_survival_time(m: int, n: int, monsters: List[Tuple[int,int]], start: Tuple[int,int]) -> int:
    INF = 10**9
    Tm = [[INF]*n for _ in range(m)]
    q = deque()
    for x,y in monsters:
        if 0 <= x < m and 0 <= y < n:
            Tm[x][y] = 0
            q.append((x,y))
    # multi-source BFS for monsters
    dirs = [(-1,0),(1,0),(0,-1),(0,1)]
    while q:
        x,y = q.popleft()
        for dx,dy in dirs:
            nx,ny = x+dx, y+dy
            if 0<=nx<m and 0<=ny<n and Tm[nx][ny] > Tm[x][y] + 1:
                Tm[nx][ny] = Tm[x][y] + 1
                q.append((nx,ny))
    sx, sy = start
    if not (0 <= sx < m and 0 <= sy < n):
        return -1
    if Tm[sx][sy] == 0:
        return -1
    # BFS for player
    Tp = [[-1]*n for _ in range(m)]
    Tp[sx][sy] = 0
    q.append((sx,sy))
    best = -1
    while q:
        x,y = q.popleft()
        t = Tp[x][y]
        if t < Tm[x][y]:
            if t > best:
                best = t
        for dx,dy in dirs:
            nx,ny = x+dx, y+dy
            if 0<=nx<m and 0<=ny<n and Tp[nx][ny] == -1:
                nt = t + 1
                # can move there, but only counts as safe if nt < Tm[nx][ny]
                Tp[nx][ny] = nt
                q.append((nx,ny))
    return best

if __name__ == "__main__":
    m,n = 5,5
    monsters = [(0,0),(4,4)]
    start=(2,2)
    print(max_survival_time(m,n,monsters,start))

File: learnLeetCode_remote/Day29/Part2/monster_survival.cpp
#include <bits/stdc++.h>
using namespace std;

int max_survival_time(int m, int n, const vector<pair<int,int>>& monsters, pair<int,int> start){
    const int INF = 1e9;
    vector<vector<int>> Tm(m, vector<int>(n, INF));
    deque<pair<int,int>> q;
    for(auto [x,y]: monsters){ if(0<=x&&x<m&&0<=y&&y<n){ Tm[x][y]=0; q.emplace_back(x,y);} }
    int dirs[4][2]={{-1,0},{1,0},{0,-1},{0,1}};
    while(!q.empty()){
        auto [x,y]=q.front(); q.pop_front();
        for(auto &d: dirs){ int nx=x+d[0], ny=y+d[1]; if(0<=nx&&nx<m&&0<=ny&&ny<n && Tm[nx][ny]>Tm[x][y]+1){ Tm[nx][ny]=Tm[x][y]+1; q.emplace_back(nx,ny);} }
    }
    int sx=start.first, sy=start.second; if(!(0<=sx&&sx<m&&0<=sy&&sy<n)) return -1; if(Tm[sx][sy]==0) return -1;
    vector<vector<int>> Tp(m, vector<int>(n, -1)); Tp[sx][sy]=0; q.emplace_back(sx,sy);
    int best=-1;
    while(!q.empty()){
        auto [x,y]=q.front(); q.pop_front(); int t=Tp[x][y];
        if(t < Tm[x][y]) best=max(best,t);
        for(auto &d: dirs){ int nx=x+d[0], ny=y+d[1]; if(0<=nx&&nx<m&&0<=ny&&ny<n && Tp[nx][ny]==-1){ Tp[nx][ny]=t+1; q.emplace_back(nx,ny);} }
    }
    return best;
}

int main(){
    int m=5,n=5; vector<pair<int,int>> monsters={{0,0},{4,4}}; pair<int,int> start={2,2};
    cout<<max_survival_time(m,n,monsters,start)<<"\n";
    return 0;
}
